### PR TITLE
estebanreyl/Unified Azure Utilities

### PR DESCRIPTION
**Pull request by estebanreyl**
Adds a singleton client for managing Azure credentials/resources internally. This helps facilitate production of further azure connectivity and extends on many of the existing methods adding optimized async support where possible. The client includes the following features:

- Acquire azureaccount
- Acquire ContainerRegistry/ResourceGroup management clients
- Validate sign in
- Get Azure Container Registries (Optional scoping by Subscription and ResourceGroup 
- Get Azure Subscriptions
- Get Azure Resource Groups (Optional scoping by Subscription)
- Get Session Credentials

### CODE REVIEW

File: utils/azureUtilityManager.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    //This function has to be called explicitly before using the singleton.
+    public setAccount(azureAccount: AzureAccount): void {
+        this.azureAccount = azureAccount;
+    }
+
+    //GETTERS
+    public getAccount(): AzureAccount {
+        if (this.azureAccount) { return this.azureAccount; }
+        throw new Error(('Azure account is not present, you may have forgotten to call setAccount'));
```
Comment (StephenWeatherford) [inline utils/azureUtilityManager.ts]: nit: double parentheses

Comment (StephenWeatherford) [inline utils/azureUtilityManager.ts]: Seems like you have three states:
1) Azure account extension available, setAccount() called
2) Azure account extension available, setAccount() not called
3) Azure account extension not available, setAccount() can't be called, getAccount() will throw if called, but no way for caller to know this

File: utils/azureUtilityManager.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        if (session) {
+            return session.credentials;
+        }
+
+        throw new Error(`Failed to get credentials, tenant ${tenantId} not found.`);
+    }
+
+    //CHECKS
+    //Provides a unified check for login that should be called once before using the rest of the singletons capabilities
+    public async isLoggedIn(): Promise<boolean> {
```
Comment (StephenWeatherford) [inline utils/azureUtilityManager.ts]: Let's keep the concept count the same by naming it waitForLogin like AzureAccount.waitForLogin

File: utils/azureUtilityManager.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        } else {
+            //Get all registries for all subscriptions
+            const subs: SubscriptionModels.Subscription[] = this.getFilteredSubscriptionList();
+            const subPool = new AsyncPool(MAX_CONCURRENT_SUBSCRIPTON_REQUESTS);
+
+            for (let sub of subs) {
+                subPool.addTask(async () => {
+                    const client = this.getContainerRegistryManagementClient(sub);
+                    let subscriptionRegistries: ContainerModels.Registry[] = await client.registries.list();
+                    registries = registries.concat(subscriptionRegistries);
```
Comment (StephenWeatherford) [inline utils/azureUtilityManager.ts]: registries.push(subscriptionRegistries);

File: utils/azureUtilityManager.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    public getAccount(): AzureAccount {
+        if (this.azureAccount) { return this.azureAccount; }
+        throw new Error(('Azure account is not present, you may have forgotten to call setAccount'));
+    }
+
+    public getFilteredSubscriptionList(): SubscriptionModels.Subscription[] {
+        return this.getAccount().filters.map<SubscriptionModels.Subscription>(filter => {
+            return {
+                id: filter.subscription.id,
+                session: filter.session,
```
Comment (StephenWeatherford) [inline utils/azureUtilityManager.ts]: Are you planning on using session (which isn't in SubscriptionModels.Subscription)?

**Comment by StephenWeatherford** [comment]
Sorry for the delay, see minor comments.

**Comment by estebanreyl** [comment]
No worries, thanks for the comments! I've updated accordingly

**Pull request merged**